### PR TITLE
NHWC support when running with MKL-DNN

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

mainly for NHWC support when running with MKL-DNN, summary changes:
(1) Add `heapFormat` (NHWC or NCHW) for models, so when one layer uses 4 dims, it can get its real format which may be not same with format from previous layer.

(2) Add `DataFormat` for DNN layers according to BLAS layers. E.g. Blas MaxPooling has DataFormat in its construct parameters, so add the same param for DNN MaxPooling.

(3) Refactor mkldnn.Linear to support NHWC input

(4) Deprecate  `DropOut` for MKLDNN, use `BlasWrapper(nn.DropOut)` if needed.

(5) Delete some constraints for NHWC, and now we can run models with NHWC input under MKLDNN backend

## How was this patch tested?

unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

**Comment by zhangxiaoli73** [comment]
jenkins pass: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/1118/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala [hunk 1]
```diff
@@ -409,30 +422,66 @@ class DnnGraph(
...
+    var heapFormat: Int = -1
+    inputs.foreach(m => {
+      if (m.shape.length == 4) {
+        return inputs(0).layout
+      }
+    })
+
+    @inline
+    def transferFormat(format: DataFormat): Int = {
+      if (format == DataFormat.NHWC) Memory.Format.nhwc else Memory.Format.nchw
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: Does it mean we have no `ntc` and `tnc`.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRToDnn.scala [hunk 2]
```diff
@@ -115,19 +115,54 @@ private[bigdl] class IRToDnn extends ConvertBase[IRElement[Float], Module[Float]
...
+    } else {
+      // special process for NHWC
+      require(t.nGroup == 1, "Only support nGroup is 1 for NHWC")
+      val layer = ReflectionUtils.reflectFromIR(node, Class.forName(prefix + "SpatialConvolution"))
+      val p = layer.parameters()
+      val weight = p._1(0)
+      val gradWeight = p._2(0)
+
+      val weightSize = weight.size() // for nchw
+      val newSize = Array(weightSize(2), weightSize(3), weightSize(1), weightSize(0))// for nhwc
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRToDnn.scala]: If the data format is `NHWC`,  the weight shape of conv is also `iohw`. Is it right ?

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/intermediate/IRconvertSpec.scala [hunk 3]
```diff
@@ -91,6 +92,144 @@ class IRconvertSpec extends BigDLSpecHelper {
...
+//    val out1 = blas.forward(input)
+//    val out2 = dnn.forward(input)
+//    Equivalent.nearequals(out1.toTensor[Float], out2.toTensor[Float], 1e-5) should be (true)
+//
+//    val gradOutput = Tensor[Float]().resizeAs(out1.toTensor[Float])
+//
+//    val gradInputBlas = blas.backward(input, gradOutput)
+//    val gradInputDnn = dnn.backward(input, gradOutput)
+//    Equivalent.nearequals(gradInputDnn.toTensor, gradInputBlas.toTensor, 1e-4) should be (true)
+//    System.clearProperty("bigdl.engineType")
```
Comment (i8run) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/intermediate/IRconvertSpec.scala]: If it's not used, please delete the comments.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala]: For 3 dims, Blas layer only supports `ntc` , so it's no need to distinguish `ntc` and `tnc`.
But for 4 dims, there are two formats `nchw` and `nhwc` in some layers, so I have to iterate through all layers to find the real format.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRToDnn.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRToDnn.scala]: Yes, weight shape of blas is `hwio`. But for dnn conv, it can still be `oihw` as input and weight format are `any`, and `reorder` will be done.

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/intermediate/IRconvertSpec.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/intermediate/IRconvertSpec.scala]: those test codes will be opened after mengceng's code merged.

**Comment by zhangxiaoli73** [comment]
jenkins pass:
http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/1124/
